File: core/src/main/java/xtdb/vector/ValueBox.java
package xtdb.vector;

import clojure.lang.Keyword;
import org.jetbrains.annotations.Nullable;

import java.nio.ByteBuffer;

public class ValueBox implements IValueWriter, IValueReader {
    private static final Keyword NULL_LEG = Keyword.intern("null");
    private static final Keyword ABSENT_LEG = Keyword.intern("absent");

    private Keyword leg;

    private long prim;
    private Object obj;

    @Override
    public Keyword getLeg() {
        return leg;
    }

    @Override
    public boolean isNull() {
        return leg == NULL_LEG;
    }

    @Override
    public boolean readBoolean() {
        return prim != 0;
    }

    @Override
    public byte readByte() {
        return (byte) prim;
    }

    @Override
    public short readShort() {
        return (short) prim;
    }

    @Override
    public int readInt() {
        return (int) prim;
    }

    @Override
    public long readLong() {
        return prim;
    }

    @Override
    public float readFloat() {
        return (float) readDouble();
    }

    @Override
    public double readDouble() {
        return Double.longBitsToDouble(prim);
    }

    @Override
    public ByteBuffer readBytes() {
        return ((ByteBuffer) obj);
    }

    @Override
    public Object readObject() {
        return obj;
    }

    @Override
    public void writeNull() {
        obj = null;
    }

    @Override
    public void writeBoolean(boolean v) {
        this.prim = v ? 1 : 0;
    }

    @Override
    public void writeByte(byte v) {
        this.prim = v;
    }

    @Override
    public void writeShort(short v) {
        this.prim = v;
    }

    @Override
    public void writeInt(int v) {
        this.prim = v;
    }

    @Override
    public void writeLong(long v) {
        this.prim = v;
    }

    @Override
    public void writeFloat(float v) {
        writeDouble(v);
    }

    @Override
    public void writeDouble(double v) {
        this.prim = Double.doubleToLongBits(v);
    }

    @Override
    public void writeBytes(ByteBuffer v) {
        this.obj = v;
    }

    @Override
    public void writeObject(@Nullable Object obj) {
        this.obj = obj;
    }

    @Override
    public IValueWriter legWriter(Keyword leg) {
        return new BoxWriter() {
            @Override
            IValueWriter box() {
                ValueBox.this.leg = leg;
                return ValueBox.this;
            }
        };
    }

}

File: core/src/main/kotlin/xtdb/arrow/ValueBox.kt
package xtdb.arrow

import java.lang.Double.*
import java.nio.ByteBuffer

private const val NULL_LEG = "null"

class ValueBox : ValueReader {
    override var leg: String? = null
        private set

    private var prim: Long = 0
    private var obj: Any? = null

    override val isNull: Boolean
        get() = leg === NULL_LEG

    override fun readBoolean() = prim != 0L
    override fun readByte() = prim.toByte()
    override fun readShort() = prim.toShort()
    override fun readInt() = prim.toInt()
    override fun readLong() = prim
    override fun readFloat() = readDouble().toFloat()
    override fun readDouble() = longBitsToDouble(prim)
    override fun readBytes(): ByteBuffer = obj as ByteBuffer
    override fun readObject(): Any? = obj

    fun writeNull() {
        leg = NULL_LEG
        obj = null
    }

    @JvmOverloads
    fun writeBoolean(leg: String? = null, v: Boolean) = writeLong(leg, (if (v) 1 else 0).toLong())

    @JvmOverloads
    fun writeByte(leg: String? = null, v: Byte) = writeLong(leg, v.toLong())

    @JvmOverloads
    fun writeShort(leg: String? = null, v: Short) = writeLong(leg, v.toLong())

    @JvmOverloads
    fun writeInt(leg: String? = null, v: Int) = writeLong(leg, v.toLong())

    @JvmOverloads
    fun writeLong(leg: String? = null, v: Long) {
        this.leg = leg
        this.prim = v
    }

    @JvmOverloads
    fun writeFloat(leg: String? = null, v: Float) = writeDouble(leg, v.toDouble())

    @JvmOverloads
    fun writeDouble(leg: String? = null, v: Double) {
        this.leg = leg
        this.prim = doubleToLongBits(v)
    }

    @JvmOverloads
    fun writeBytes(leg: String? = null, v: ByteBuffer) {
        this.leg = leg
        this.obj = v
    }

    @JvmOverloads
    fun writeObject(leg: String? = null, obj: Any?) {
        this.leg = leg
        this.obj = obj
    }
}
